core/src/main/scala/zio/schema/Schema.scala

```diff
@@ -27,6 +27,410 @@ object Schema {
   sealed case class Tuple[A, B](left: Schema[A], right: Schema[B])       extends Schema[(A, B)]
   final case class EitherSchema[A, B](left: Schema[A], right: Schema[B]) extends Schema[Either[A, B]]
 
+  sealed trait CaseClass[Z] extends Schema[Z] {
+    def toRecord: Record
+  }
+  case class CaseClass1[A, Z](field: (String, Schema[A]), construct: A => Z, extract: Z => Tuple1[A])
+      extends CaseClass[Z] {
+    override def toRecord: Record = Record(Map(field))
+  }
+  case class CaseClass2[A1, A2, Z](
+    field1: (String, Schema[A1]),
+    field2: (String, Schema[A2]),
+    construct: (A1, A2) => Z,
+    extract: Z => (A1, A2)
+  ) extends CaseClass[Z] {
+    override def toRecord: Record = Record(Map(field1, field2))
+  }
+
+  case class CaseClass3[A1, A2, A3, Z](
```

Comment:
Here and elsewhere, I'd use:

```scala
final case class
```

Since it does not make sense to override case classes.